pyanaconda/installclass.py

```diff
@@ -175,114 +178,105 @@ def setNetworkOnbootDefault(self, ksdata):
-            log.warning("module import of %s failed: %s", mainName, sys.exc_info()[0])
-
-    # sort by sort order first, then by install class name
-    lst.sort(key=lambda x: (x[1], x[0][0]))
-    for (item, _) in lst:
-        if showHidden:
-            allClasses_hidden += [item]
-        else:
-            allClasses += [item]
-
-    if showHidden:
-        return allClasses_hidden
-    else:
-        return allClasses
-
-def getBaseInstallClass():
-    # figure out what installclass we should base on.
-    allavail = availableClasses(showHidden=True)
-    avail = availableClasses(showHidden=False)
-
-    if len(avail) == 1:
-        (cname, cobject) = avail[0]
-        log.info("using only installclass %s", cname)
-    elif len(allavail) == 1:
-        (cname, cobject) = allavail[0]
-        log.info("using only installclass %s", cname)
-
-    # Use the highest priority install class if more than one found.
-    elif len(avail) > 1:
-        (cname, cobject) = avail.pop()
-        log.info('%s is the highest priority installclass, using it', cname)
-    elif len(allavail) > 1:
-        (cname, cobject) = allavail.pop()
-        log.info('%s is the highest priority installclass, using it', cname)
-    else:
-        raise RuntimeError("Unable to find an install class to use!!!")
-
-    return cobject
-
-baseclass = getBaseInstallClass()
-
-# we need to be able to differentiate between this and custom
-class DefaultInstall(baseclass):
+
+class InstallClassFactory(object):
+    """Class to get the install class instance."""
+
     def __init__(self):
-        baseclass.__init__(self)
+        self._classes = []
+        self._visible_classes = []
+        self._paths = []
+
+    def get_base_install_class(self):
+        """Return the instance of the best found install class."""
+        if self.visible_classes:
+            install_class = self.visible_classes[0]
+            log.info("Using a visible install class %s.",
+                     self._get_class_description(install_class))
+
+        elif self.classes:
+            install_class = self.classes[0]
+            log.info("Using a hidden install class %s.",
+                     self._get_class_description(install_class))
+
+        else:
+            raise RuntimeError("Unable to find an install class to use.")
+
+        return install_class()
+
+    @property
+    def paths(self):
+        """Return paths where to look for install classes."""
+        if not self._paths:
+            path = []
+
+            if "ANACONDA_INSTALL_CLASSES" in os.environ:
+                path += os.environ["ANACONDA_INSTALL_CLASSES"].split(":")
+
+            path += [
+                "installclasses",
+                "/tmp/updates/pyanaconda/installclasses",
+                "/tmp/product/pyanaconda/installclasses",
+                "%s/pyanaconda/installclasses" % get_python_lib(plat_specific=1)
+            ]
+
+            self._paths = list(filter(lambda d: os.access(d, os.R_OK), path))
```

Comment:
What about hiding this implementation to some function? For example `find_all_installclass_paths()`?

It would be easier to read:

```
if not self._paths:
    self.find_all_installclass_paths()
return self._paths
```

It is the same as you have for ``classes`` property.